### PR TITLE
Optimize count agg expr with null column statistics

### LINKED ISSUE #904: Optimize `count(col)` using table statistics
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

If both the number of rows and the number of nulls are available as statistic, we can compute the value based on statistics only when reading.
`row_count - null_count(row)`

**Describe the solution you'd like**
Add an optimization rule that matches this count expression with column and use the statistics to replace it with the value.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by matthewmturner**
# Which issue does this PR close?

Closes #904 

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by matthewmturner** [comment]
@Dandandan @rdettai FYI - I took a first cut at this.  Still reviewing on my end - but wanted to at least open the PR in case either of you had specific thoughts.

File: datafusion/src/physical_optimizer/aggregate_statistics.rs [hunk 1]
```diff
@@ -153,6 +159,37 @@ fn take_optimizable_count(
...
+                .downcast_ref::<expressions::Column>()
+            {
+                if let ColumnStatistics {
+                    null_count: Some(val),
+                    ..
+                } = &col_stats[col_expr.index()]
+                {
+                    return Some((
+                        ScalarValue::UInt64(Some((num_rows - val) as u64)),
+                        "COUNT(Uint8(1))".to_string(),
```
Comment (Dandandan) [inline datafusion/src/physical_optimizer/aggregate_statistics.rs:184]:
```suggestion
                        "COUNT(UInt8(1))".to_string(),
```

**Comment by Dandandan** [comment]
I think the approach looks good @matthewmturner 

Code could be simplified a bit or take more cases into account (as mentioned in TODO)

**Comment by matthewmturner** [comment]
> I think the approach looks good @matthewmturner
> 
> Code could be simplified a bit or take more cases into account (as mentioned in TODO)

To confirm, are you referring to the "Optimize with exprs other than Column"?

File: datafusion/src/physical_optimizer/aggregate_statistics.rs (hunk 1 above)
Comment (rdettai) [reply datafusion/src/physical_optimizer/aggregate_statistics.rs]: Shouldn't this be the column name instead of `UInt8`?

**Comment by rdettai** [comment]
It would be very reassuring if these optimizations came with a more systematic testing routine. Currently we have a test here: https://github.com/apache/arrow-datafusion/blob/558d8ecfe2d1b0737de1e2548c5fbedc0ea17ada/datafusion/tests/sql.rs#L3020

that ensures that the count optimizer rule kicks in properly with explain optimize. What do you think about creating a separate integration test file called `optimizers.rs` that runs SQL queries and checks if they were optimized as expected?

File: datafusion/src/physical_optimizer/aggregate_statistics.rs (hunk 1 above)
Comment (Dandandan) [reply datafusion/src/physical_optimizer/aggregate_statistics.rs]: I think it should be the column name here indeed for it to work.
The current way hash aggregates work is by accessing the value of the aggregate in a projection by the name generated by the expression in the aggregate (e.g. `COUNT(UInt8(1))` for `count(*)` but `COUNT(col)` for a non null count. So by rewriting and giving it the same name, this avoids computing the aggregate while still being compatible with the usage of the aggregate in the projection.

**Comment by matthewmturner** [comment]
> It would be very reassuring if these optimizations came with a more systematic testing routine. I agree that unit tests in this precise scenario are not very powerful. Currently we have a test here:
> 
> https://github.com/apache/arrow-datafusion/blob/558d8ecfe2d1b0737de1e2548c5fbedc0ea17ada/datafusion/tests/sql.rs#L3020
> 
> that ensures that the count optimizer rule kicks in properly with `EXPLAIN [ANALYZE]`. What do you think about creating a separate integration test file called `optimizers.rs` that runs SQL queries and checks if they were optimized as expected?

I'm happy to do this.

Do you think better to make that file on a separate PR or maybe i could just add it for these count optimizations for now and a separate PR for the other optimizations?

**Comment by rdettai** [comment]
Thanks for your motivation! In this PR only tests for this feature are required.

**Comment by matthewmturner** [comment]
@rdettai @Dandandan - I got this passing CI.  Will start looking into tests I can do for this feature now.  I'm going to need to study the structure a bit though to figure out the right way to do it - if you have anything specific in mind let me know.

**Comment by matthewmturner** [comment]
@Dandandan @rdettai FYI I added a few tests.

File: datafusion/src/physical_optimizer/aggregate_statistics.rs [hunk 2]
```diff
@@ -276,6 +336,35 @@ mod tests {
...
     }
 
+    /// Checks that the count optimization was applied and we still get the right result
+    async fn assert_count_with_nulls_optim_success(
+        plan: HashAggregateExec,
+    ) -> Result<()> {
+        let conf = ExecutionConfig::new();
+        let optimized = AggregateStatistics::new().optimize(Arc::new(plan), &conf)?;
+
+        assert!(optimized.as_any().is::<ProjectionExec>());
```
Comment (Dandandan) [inline datafusion/src/physical_optimizer/aggregate_statistics.rs:346]: Maybe add a comment here that the added `ProjectionExec` is a sign the optimization was applied.

**Comment by matthewmturner** [comment]
@Dandandan @rdettai can you provide some color on why we have different return expressions for the `take_optimizable_count` and `take_optimizable_count_with_nulls`?

Specifically, the expression for `take_optimizable_count` is `"COUNT(Uint8(1))"` and for `take_optimizable_count_with_nulls` is `"COUNT({})", col_expr.name()"`

**Comment by houqp** [comment]
I think the following two specs are related to the discussion here:

https://arrow.apache.org/datafusion/specification/invariants.html#physical-schema-is-invariant-under-physical-optimization
https://arrow.apache.org/datafusion/specification/output-field-name-semantic.html

Note that the existing specs are not set in stone, so we can propose changes to them if we can back them with good reasons.

**Comment by matthewmturner** [comment]
@Dandandan @rdettai do you have any thoughts on this?

**Comment by matthewmturner** [comment]
@alamb do you have any thoughts on this PR / how to move forward?

**Comment by alamb** [comment]
Sorry for the delay @matthewmturner  -- let me look more carefully at this PR. Note I think @rdettai  is out this week and next which may explain the slow response. I have no such excuse however :)

**Review by alamb — COMMENTED** [review]
I ran out of time to do a more thorough review, but I left some hints -- hopefully they make sense @matthewmturner ?

**Comment by matthewmturner** [comment]
@alamb no apology necessary, thank you for the feedback. i will look into your comments.

File: datafusion/src/physical_optimizer/aggregate_statistics.rs [hunk 3]
```diff
@@ -250,38 +289,41 @@ mod tests {
...
-    async fn assert_count_optim_success(plan: HashAggregateExec) -> Result<()> {
+    async fn assert_count_optim_success(
+        plan: HashAggregateExec,
+        nulls: bool,
+    ) -> Result<()> {
         let conf = ExecutionConfig::new();
         let optimized = AggregateStatistics::new().optimize(Arc::new(plan), &conf)?;
 
+        let (col, count) = match nulls {
+            false => (Field::new("COUNT(Uint8(1))", DataType::UInt64, false), 3),
```
[the remark this answers is not in the document]
Comment (matthewmturner) [reply datafusion/src/physical_optimizer/aggregate_statistics.rs:300]: i think this is just a case of a poorly named variable.  it should really be something like `col_count` i.e. if the test is on a table count or column count.  when using datafusion-cli on this branch i get the same output as what you showed. i will update.

File: datafusion/src/physical_optimizer/aggregate_statistics.rs [hunk 4]
```diff
@@ -153,6 +159,39 @@ fn take_optimizable_count(
...
+fn take_optimizable_count_with_nulls(
+    agg_expr: &dyn AggregateExpr,
+    stats: &Statistics,
+) -> Option<(ScalarValue, String)> {
+    if let (Some(num_rows), Some(col_stats), Some(casted_expr)) = (
+        stats.num_rows,
+        &stats.column_statistics,
+        agg_expr.as_any().downcast_ref::<expressions::Count>(),
+    ) {
+        if casted_expr.expressions().len() == 1 {
```
[the remark this answers is not in the document]
Comment (matthewmturner) [reply datafusion/src/physical_optimizer/aggregate_statistics.rs:172]: My understanding is that `COUNT(*)` doesnt need to have a separate handler for nulls - assuming we expect same behavior as psql.  For example in psql when i do the following:
```
postgres=# create table foo as select * from (values (1,NULL),(NULL,2),(3,3)) as sq;
SELECT 3
postgres=# select * from foo;
 column1 | column2
---------+---------
       1       |
                |       2
       3       |       3
(3 rows)

postgres=# select count(*) from foo;
 count
-------
     3
(1 row)
```
Does it make sense to reframe these optimizations as the following:
'take_optimizable_table_count' (current `take_optimizable_count`)=> comes from `COUNT(*)` and returns `num_rows`
`take_optimizable_column_count` (current `take_optimizable_count_with_nulls`) => comes from `COUNT(col)` and return `num_rows - null_count` for `col`

[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/src/physical_optimizer/aggregate_statistics.rs:172]: I think those names make more sense to me

[the remark this answers is not in the document]
Comment (matthewmturner) [reply datafusion/src/physical_optimizer/aggregate_statistics.rs:172]: Ok - ive updated.  Let me know if anything else needed.

**Review by alamb — COMMENTED** [review]
@matthewmturner  something seems to have gone wrong with the most recent merge (it has 10k lines added 🤔 )

Screen Shot 2021-11-15 at 5 12 00 PM

I am not sure what is going on here. 

Perhaps it is time to rebase this PR (I am happy to help do so) -- this PR has been hanging out for so long I want to help get it in

**Review by rdettai — APPROVED** [review]
I think this code will need some refactoring before being able to cover more cases of aggregation simplification, but this is good enough for now.

File: datafusion/src/physical_optimizer/aggregate_statistics.rs (hunk 3 above)
Comment (Dandandan) [inline datafusion/src/physical_optimizer/aggregate_statistics.rs:300]:
```suggestion
            false => (Field::new("COUNT(UInt8(1))", DataType::UInt64, false), 3),
```

**Comment by matthewmturner** [comment]
> 
> 
> I think this code will need some refactoring before being able to cover more cases of aggregation simplification, but this is good enough for now.
@rdettai are you referring to handling more than just column expressions?  if so, i can do a follow on pr for updating this and the min/max optimizations (which this was based on).

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you for sticking with it @matthewmturner !

**Pull request merged**